Add FormSpec key and mouse events

v-rob opened the pull request:
This PR allows FormSpecs to catch arbitrary keypresses and mouse events and pass them to `minetest.register_on_player_receive_fields`.  This creates two extra fields: `key_event` and `mouse_event`.  This fixes #6768, #6353, and #7451.

## Documentation from `lua_api.txt`

Special Input Events
--------------------

> Formspecs can return two different special events not originating from any element, `key_event`
> and `mouse_event`, after they have been enabled explicitly in the `options[]` element. Both
> events are a string that can be translated into a table using `minetest.explode_mouse_event`
> and `minetest.explode_key_event`. If the event is malformed (e.g. a client sending incorrect
> or malicious data), both exploders will return nil.
> 
> The tables returned by each event exploder contain the following fields:
> 
> ### `key_event`:
> * `key`: The name of the key that was pressed. A list can be found in `doc/keycodes.txt`.
>     * The keycodes are keyboard layout-dependent. The position of the W key on a QWERTY
>       keyboard differs from that of a AZERTY keyboard; thus, these are not suitable for
>       things dependent on key positions.
>     * Currently, characters like `&` which are only accessed via the Shift key and some
>       other layout-dependent key are not returned in this field.
>     * Do NOT use keycodes for custom textual input fields; they will not work for
>       non-English languages.
>     * Unknown keys will return the value "Unknown Key".
> * `mapped`: If the key is also an in-game key (for instance, the key for walking forward),
>   this field contains the keymap name (in this case, "forward" since "keymap_forward" is
>   the name for the setting for walking forward); otherwise, it is nil.
> * `pressed`: True if the key was pressed down, false if released.
> 
> `key_event` will not fire if focus is in an element that handles those keys itself. For
> instance, `button[]`s consume the spacebar and escape keys, while `field[]`s consume
> nearly all keys.
> 
> ### `mouse_event`
> * `type`: The type of the mouse event. One of "press", "release", "move", or "wheel".
> * `button`: If `type` is "press" or "release", this is the mouse button that was
>   pressed or released. One of "left", "middle", or "right". `nil` if a different event.
> * `clicks`: If `type` is "press" or "release", this is how many times the mouse
>   button was pressed in quick succession (i.e. 1 for single click, 2 for double click,
>   and 3 for triple click). Currently, it will not count higher than 3, but it may go
>   higher in the future. `nil` if a different event.
> * `form_pos`: A 2D vector that contains the mouse position in fractions of coordinates
>   from the top left corner of the formspec.
> * `elem`: The name of the currently hovered element. `nil` if not hovering over an
>   element or the element has no name.
> * `elem_pos`: A 2D vector that contains the mouse position in fractions of coordinates
>   from the top left corner of the currently hovered element. If `elem` is `nil`, this
>   contains the same value as `form_pos`.
> * `wheel`: The amount of change in the mouse wheel. It is >0 if the mouse wheel was
>   scrolled up, <0 if down, and 0 for non-wheel events. It can be a decimal value,
>   which is common with laptop touchpads.
> 
> The "move" event will not occur if the mouse event is set to `no_move`. In addition,
> the "move" event may only be sent once every 200 milliseconds in `all` to prevent
> it from bogging down network traffic.

## Todo

This PR is Ready for Review, but testing is needed from other people to make sure the key changes work.

## How to test

The `Real Coordinates` tab in Devtest `/test_formsec` has everything enabled, and `Styling` has `no_move`.

v-rob committed:
Add FormSpec key and mouse events

ClobberXD commented:
> Neither of these will return a value if focus is on any element in a FormSpec that requires key input, e.g. field, textarea, textlist, table, etc.

I sorta understand the reason behind this decision, but I think it'd actually be useful if keypress events are captured even when focussing on a text-based formspec element. One use-case I can immediately think of is the Online tab - the username or password field getting reset mid-way while typing is the most annoying thing ever. Using these events, the client can cache whatever the user types, upon each keypress, and can then restore the field's data once the serverlist has been populated.

v-rob commented:
The keypresses acting that way was not a decision I made.  The event handler isn't fired inside these for some reason -- you'd probably have to edit the code for these elements to get it to work.

Another thing is that this usecase actually isn't possible because combinations of keys, e.g. `shift + a` to get capital A cannot be determined through the raw keypresses themselves.

I think that textareas, fields, and such should have their own code to fix this themselves anyway since that would be a hacky fix.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3138,6 +3138,61 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
+			SET_KEY_MAPPED("keymap_quicktune_next");
+			SET_KEY_MAPPED("keymap_quicktune_prev");
+			SET_KEY_MAPPED("keymap_quicktune_inc");
+			SET_KEY_MAPPED("keymap_quicktune_dec");
```
Can outdate very quickly. Consider adding a function to `Settings` which returns an array of these `keymap_*` settings.
EDIT: You can also abort as soon one setting matches the key value.

v-rob replied:
Yes, I thought that there should be a function, but I don't see how to get a list of all settings, so it might have to be another hardcoded list.  Aborting is a good idea, I'll do that.

As for EDIT2, these don't *do* anything on their own.  The name is returned when a modder presses one of these keys.

SmallJoker commented on doc/keycodes.txt:
```diff
@@ -0,0 +1,162 @@
+When choosing keys, try not to use arcane keys; instead, choose keys that are
+one every or most keyboards.
+
+* `KEY_LBUTTON`          = Left mouse button
```
Minetest does not control this enum, so it would best to only refer to the Irrlicht docs and/or source.

rubenwardy replied:
Irrlicht is unlikely to receive any more updates, so this is probably fine

v-rob replied:
I think ease for modders is necessary.  And if Irrlicht receives an update, we'll have to patch it anyway because it would break any mods that rely on the current system.

v-rob replied:
Actually, the names returned in `name` are returned in the `KeyPress` class which generates the names itself, so even if Irrlicht updates, the key names will stay the same.  So this isn't a problem.

SmallJoker commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -22,6 +22,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include <utility>
 #include <stack>
 
+#include "client/keycode.h"
```
Only needed in `guiFormSpecMenu.cpp`. Here you can use the forward declaration `class KeyPress;` for compiling speed reasons.

v-rob replied:
No idea how that works.  C++ is too weird and bloated.  Do I place `class KeyPress` anywhere in the file, or as part of the variable type?

DS-Minetest replied:
Afaik, all you have to do is to write the class header:
```cpp
// include everything

class KeyPress;

// declare other class that has a member variable that needs the class KeyPress
```
It's just like the function forward declaration which is done in header files.

SmallJoker replied:
You can replace this commented line with `class KeyPress;`. Then undo the removal of `#include "client/keycode.h"` in the cpp file.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -3928,6 +3929,14 @@ Call these functions only at load time!
+          inside the form. As opposed to `key_event`, this field returns keys
+          only if they are mapped to an in-game key. The keycode will be the
+          setting name found in `minetest.conf`. This will not return any keys
+          if focus is inside an element that receives key input.
```
What's the use-case of knowing the setting name for the key?

ClobberXD replied:
This (probably) allows performing different actions, for different keys being pressed.

DS-Minetest replied:
- Sneak-clicking something.
- Pressing the drop button to make something drop. (This makes me think, that it would be nice to have the cursor pos, too.
- Controlling something with the real forward, backward, and so on buttons.

All these things wouldn't work well with different keyboard layouts.

SmallJoker replied:
`key_mapped` values are `keymap_forward`, `keymap_toggle_chat` and so on (see L3146).
You know that some key is mapped, but not which one.

DS-Minetest replied:
As far as I understand the code [here](https://github.com/minetest/minetest/pull/8679/files#diff-074850a88796e13dad12dd7ae1a8d9f7R3143) correctly, it will be eg. `fields = {key_event = "KEY_KEY_D", key_mapped = "keymap_right", ...}`. And ergo, you know the mapped key.

v-rob replied:
*Possible case:*

* Player 1's in-game forward key is `KEY_W`.
* Player 2's in-game forward key is `KEY_UP`.

So, if some code wants to get the in-game forward key, this code
```lua
if fields.key_event == "KEY_W" then
    do_stuff()
end
```
will only work for Player 1.

On the other hand, try this code:
```lua
if fields.key_mapped == "keymap_forward" then
    do_stuff()
end
```
This will work for all players.

ClobberXD replied:
Ah I mixed it up. The usecases are valid nonetheless. Here's another one: Change key menu. I don't know how the current change keys menu captures keypresses, but this PR would allow for one such menu to be accessible and modifiable in-game.

LoneWolfHT commented:
>Using these events, the client can cache whatever the user types, upon each keypress, and can then restore the field's data once the serverlist has been populated.

Might be better to just add a way for modders to cache whatever text is in the fields before they refresh the formspec

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3453,22 +3508,21 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
+
+			if (m_client != NULL && (kp == getKeySetting("keymap_screenshot"))) {
+				m_client->makeScreenshot();
+			}
```
Oh, formspec screenshots, nice!

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3453,22 +3508,21 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
+				return true;
+			}
+
+			if (m_client != NULL && (kp == getKeySetting("keymap_screenshot"))) {
```
```suggestion
			if (m_client && kp == getKeySetting("keymap_screenshot")) {
```

ClobberXD replied:
On a side-note: Does this fix the problem of `keymap_screenshot` keypress being ignored when a formspec is open?

v-rob replied:
Is it ignored?  I can take screenshots while formspecs are open, and it is returned properly in `key_mapped`, so I don't see the problem.

ClobberXD replied:
It's ignored in the main menu. Or maybe I'm wrong about this. It's probably initialized only when a game is launched. I assumed it had something to do with formspecs, sorry. :)

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3453,22 +3508,21 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
+			current_keys_pending.key_event = kp;
+
+			if ((kp == EscapeKey) || (kp == CancelKey) ||
+					((m_client != NULL) && (kp == getKeySetting("keymap_inventory")))) {
```
```diff
-			if ((kp == EscapeKey) || (kp == CancelKey) ||
-					((m_client != NULL) && (kp == getKeySetting("keymap_inventory")))) {
+			if (kp == EscapeKey || kp == CancelKey ||
+					m_clinet && kp == getKeySetting("keymap_inventory")) {
```

DS-Minetest commented:
This PR might close #6353 and #7451.

v-rob commented:
> > Using these events, the client can cache whatever the user types, upon each keypress, and can then restore the field's data once the serverlist has been populated.
> 
> Might be better to just add a way for modders to cache whatever text is in the fields before they refresh the formspec

I was thinking that it should be automatic because data is lost when the form is regenerated when the Minetest window is resized as well.

v-rob commented:
I'm going to expand the scope of this PR.  First off, I will give many more return values that can be gotten from the string with `minetest.deserialize`.  Second, I will give events for the mouse.  Third, `key_mapped` will be merged into `key_event`

Return values for `key_event`:
```lua
{
    name = <key name>,            -- Key name, e.g. "KEY_UP"
    pressed = <bool>,             -- true = pressed, false = released
    control_down = <bool>,        -- Is control held down during this keypress?
    shift_down = <bool>,          -- Is shift held down during this keypress?
    mapped = <mapped name>/false, -- minetest.conf keymap name or false if none
}
```
Return values for `mouse_event`:
```lua
{
    type = "click"/"release"/"doubleclick"/"tripleclick"/"move"/"wheel",
    left_down = <bool>,           -- Is left mouse button down?
    right_down = <bool>,          -- Is right mouse button down?
    middle_down = <bool>,         -- Is middle mouse button down?
    control_down = <bool>,        -- Is control key down?
    shift_down = <bool>,          -- Is shift key down?
    x = <number>,                 -- X mouse pos in coordinates, not pixels
    y = <number>,                 -- Y mouse pos in coordinates, not pixels
    wheel = ?,                    -- Probably a number
}
```

DS-Minetest commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -34,6 +34,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 class InventoryManager;
 class ISimpleTextureSource;
 class Client;
+class KeyPress
```
```suggestion
class KeyPress;
```

red-001 commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3138,6 +3140,120 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
+				else if (event.Event == e)	\
+					field_type = t;
+
+			if (event.Event == EMIE_LMOUSE_PRESSED_DOWN)
```
why doesn't this also use `SET_MOUSE_EVENT`? Is this meant to make it more obvious what it does?

v-rob replied:
`SET_MOUSE_EVENT` uses the `else if` statement, whereas this first one needs the `if` statement, so it has to be different.

red-001 commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3138,6 +3140,120 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
+
+			std::string field_type = "error";
+
+			#define SET_MOUSE_EVENT(e, t)	\
```
`DETECT_MOUSE_EVENT` might be a better name?

v-rob replied:
Probably.  I need to rebase, so I might as well do this anyway.

red-001 commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3138,6 +3140,120 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
+			std::string field_name = current_keys_pending.key_press.sym();
+			std::string field_mapped = "nil";
+
+			#define SET_KEY_MAPPED(m)											\
```
Same comments on this as above.

kilbith commented:
I think that feature rocks and should be added to the formspec priority list for the next release.

luk3yx commented:
I suggest against using `minetest.deserialize` on client-provided input, it gives clients a way to crash/hang servers (you can still break things without access to any variables). You could possibly use JSON instead (and recommend validating input) or create a custom format and a `minetest.explode_key_event`.

paramat commented:
@v-rob are you still working on your PRs? I see a lot of yours have not progressed for several months.

v-rob commented:
Funny, I got a notification about paramat commenting, but there's nothing here from him.

Anyhoo, this PR has been recently rebased and just now, I made the `mapped` key of `key_event` use a new function of `Settings` to get all `keymap_` settings.  It also removes the unnecessary `keymap_` prefix,  I recently realized that `mapped` isn't strictly necessary because the following pieces of code are equivalent:

```lua
if key_event.name == minetest.settings:get("keymap_forward") then
    -- do stuff
end
```
```lua
if key_event.mapped == "forward" then
    -- do stuff
end
```
`mapped` might be slightly faster, but the real reason I want to keep it is because it makes it easier for modders (simpler and doesn't have `keymap_`) and more likely for them to not hardcode key values to `KEY_KEY_W` or something like that.

paramat commented:
> Funny, I got a notification about paramat commenting, but there's nothing here from him.

Yes, i made a comment about this being neglected, based on your comment from July, but then saw the recent force-pushes, so it is not neglected. Deleted my comment.

luk3yx commented on doc/lua_api.txt:
```diff
@@ -2360,12 +2360,69 @@ Elements
+
+Formspecs return two different special events not originating from any element:
+`key_event` and `mouse_event`. Both events give JSON that can be translated into
+a table using `minetest.parse_json`.
```
Couldn't the client send invalid JSON (or just send `null`) and crash the server?

v-rob replied:
Does invalid JSON crash the server, or does the function just return `nil` if it's invalid?  But yeah, I should recommend making sure it gives correct output.

I don't feel like adding a new `explode_key_event` and an `explode_mouse_event` since they seem kind of redundant, but if I have to, I suppose I will.

luk3yx replied:
Invalid JSON [returns `nil` as well as logging an error](https://github.com/minetest/minetest/blob/9284313d17ba906de449c6d39ef3ddacb6c73791/doc/lua_api.txt#L5135).

v-rob replied:
Well, that settles it.  I'll work on `explode_key_event` and `explode_mouse_event` then.

v-rob commented:
I've finished the `minetest.explode_key_event` and `minetest.explode_mouse_event`, so everything is ready for review.

sorcerykid commented:
Of all the pending formspec PRs, I'm most excited about this one. With this feature it will be possible for players on my server to navigate the in-game GPS map using the arrow keys rather than clunky buttons. And they'll even be able to click on specific areas of the map to precisely set markers for travel destinations. This is going to so rock!

Zughy commented:
This is huge, chapeau. Any news if we're going to see v-rob improvements on formspecs on 5.4?

v-rob commented:
Well, I have two PRs with two approvals and another with one approval: https://github.com/minetest/minetest/pulls/v-rob.  A few others are of interest that I expect might get some attention after 5.3 is released; this might be one of them.

BruceMustache commented:
This will be awesome. What is missing for this pull request be done?

v-rob commented:
Apparently a rebase that I didn't notice.

I really need to learn how to squash...

LandonAConway commented:
This would be very useful. Would be even more useful to have a similar feature that allows this type of stuff while not doing anything with formspecs. For example, a mod that allows players to double click then run faster that the norman 'fast speed. Or to toggle other stuff.

v-rob commented:
There's one main problem with adding this to the normal game: Keypresses must be caught using `minetest.register_globalstep`, so they can be missed.  This PR doesn't give how many keypresses are held down at once like `player:get_player_control()`, but every time the keys change and describes the change.  Since they can be missed, a player could press a key between a globalstep and no mod would ever know about it.  Some kind of `minetest.register_on_keypress` type of callback must be developed before I would say that this type of thing could be added to the actual game.  Even then, I don't know, since that callback would be called a _lot_ since players press keys all the time.

Also, I have no knowledge of any of Minetest's code beyond formspecs, so that would be tricky for me specifically.

Anyway, the core devs still need to review this first (especially since it seems to be one of my most popular PRs).

v-rob commented:
Since no core devs have reviewed yet, I decided that adding a few more useful features would be OK, especially since they probably should have been there from the start.

* `mouse_event` gives the name of the currently hovered element and the offset from its top left corner as the fields `elem` and `elem_pos`, respectively.  `x` and `y` have been renamed `form_pos`.  Both `form_pos` and `elem_pos` are tables with `x` and `y` keys.
    * The rationale for this is that, while it is possible to reverse-calculate what element is being hovered, this is error-prone, and things like containers have to be taken into account.  Doing it in C++ is much simpler, faster, and less error-prone.
* `key_event` has a new field, `pressed`, which is `true` if the key was pressed and `false` if released.  This is akin to the `mouse_event` `type` key, and I don't know why I didn't have it before.
* A problem in both of the `minetest.explode_<type>_event` functions was fixed where the usage of `string.split` would ignore empty slots.
* Made the mouse wheel work properly with all mice.

pyrollo commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3951,14 +4048,8 @@ enum ButtonEventType : u8
-			return true;
-		}
 
 		if (m_client != NULL && event.KeyInput.PressedDown &&
```
There are two "if F12 then screenshot" stuff. If you take screenshot with this version, you get screenshot twice.
This one should be removed and only the other "if F12 then screenshot" below should be kept.

pyrollo commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3968,11 +4059,21 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
+				return true;
+			}
+
+			if (m_client && kp == getKeySetting("keymap_screenshot")) {
```
This is the "if F12 then screenshot" to be kept.

pyrollo commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -40,6 +41,7 @@ class ISimpleTextureSource;
 class Client;
 class TexturePool;
 class GUIScrollContainer;
+class KeyPress;
```
Forward declaration of KeyPress is useless as it is defined in client/keycode.h (include needed for fs_key_pendig declaration).

pyrollo reviewed (dismissed):
Apart from the double screenshot issue, seems to work well. There is however something that may be annoying:

If I have a formspec performing some checks on fields before checking which event (which button) triggered the receive_fields, its behavior may be affected by these new mouse and keyboard events.

Adding a specific element to tell which events are listened to (`events[mouse,keyboard]` for example or `events[true]`), disabled by default would ensure that no mods are affected.

For future versions, this event notion could be generalized. It could be good to have an `event` special field that tells what did trigger the receive_fields : mouse, keyboard, gui, and in that last case, which element and interaction did occure.

Desour reviewed (commented):
Code looks mostly fine, I didn't test.

But as most formspecs won't use any of the new key and mouse events, there really needs to be something like a new element to enable these new events, especially since there will probably be a massive amount of mouse movement events.

Mouse events could even be bound to a new element, something like a rect that players can click and hover with their mouse.

pyrollo commented:
I had an idea about problem with filtering events. Why not take chance of this PR to introduce a new formspec event management?

Keeping `minetest.on_register_player_receive_fiedls` as it is (not sending more stuff), but have a `minetest.on_register_gui_event_handler` (or something else), returning more information (and more often) about gui event and fields.

Format could be something like:

```
{
  event = {
    cause = "mouse" / "gui" / "key" ,  -- What caused the event
    ... : Definition of the event
  }
  mouse = { x = ... , y= ..., btn1=..., btn2 =... } -- Mouse state (even for gui / key events)
  fields = { ... } -- Values of current fields ( text fields, lists, sliders, checkbox, everything holding a value, but not buttons or like)
}
```
Definition of the event could include a lot of different things:
- Gui button pressed, checkbox changed, list event, ...
- Inventory events;
- Mouse button clicked, wheel (or move?);
- Key pressed;

Many new things could be done. For example, having mouse position when clicking on an image could make it possible to make image mapping.

sorcerykid commented:
I like that idea in principle. It somewhat parallels the main menu API which has a separate hook for events.

However, I was hoping to see this PR merged sooner rather than later, as the feature set is desperately needed by a lot of modders. Also, v-rob understandably has no desire to implement additional features at this late stage. I think revamping the API is well beyond the scope of this PR, and perhaps more suitable to a discussion about a formspec replacement. Just some thoughts.

rubenwardy commented:
I don't think an event rewrite is necessary for this PR, but I do think that it's worth adding a formspec element to enable key/mouse events. This could be significant traffic

pyrollo commented:
I'll try again the next time :)

v-rob commented:
It does seem like a pretty good idea (in fact, that looks much like what Irrlicht's events look like internally), but yeah, out of scope for this PR.  A big event overhaul seems like a good thing for when we're replacing the current syntax.

I'm sorry I haven't addressed the things that need to be done yet; the refactor's been taking most of my time.  I'll try to get them addressed today.

v-rob commented:
Addressed the issues and added the `events[mouse,key]` element to turn them on.  I didn't choose `<type>_events[<bool>]` because key and mouse events are global, so having a boolean to toggle it on and off seems useless.

sorcerykid commented:
I'm not quite sure how being global relates to avoiding booleans. The `no_prepend[]` element is global and effectively a boolean too. Let's consider, for sake of argument, that eventually other global formspec options can be toggled. What is the best methodology for doing so?

I would personally favour global options either being toggled via individual elements or otherwise introducing a new element for this purpose. Perhaps something like `options[key_events=true,mouse_events=false,prepend=true]`, appearing immediately after the `size[]`, `position[]`, and `anchor[]` elements. I feel these are both cleaner solutions. But I'm open to different opinions.

v-rob commented:
My statement about booleans being useless is like this: `real_coordinates` can be toggled on and off throughout the formspec since it applies to elements individually, but prepends can only be turned off since they are already on.  In the case of this PR, saying `mouse_event[false]` would be worthless since there is no point in turning something off that's already off by default.

On the other hand, making a unified `options` element for global form options isn't a bad idea.  Saves having to make new elements for everything.  I'll do that.

pyrollo commented:
I have two minor remarks (that wont prevent from giving approval):

- Number of fixed order elements begins to grow (size > position > anchor > no_prepend > options > real_coordinates). I'm not sure it's always relevant to fix that order (it may generate unnecessary errors). Hope in the furure we will have two main types of elements: "global" elements and "other" elements. "global" elements just having to be before any "other" elements but in any order.

- Mouse move events could be filtered apart with an option. If, say, I want to control mouse wheel, I'll end up with a huge amount of unwanted mouse move events.

This may be out of this PR scope and could be done later. PR is ok enough for me to merge it and see how all that is used.

pyrollo reviewed (dismissed):
I get some exotic key mapping with my AZERTY keyboard. For example, if I press shift-4 (shift is necessary on Azerty to get the number instead of single quote), I get a KEY_OEM_7. For 6 and 8 I get KEY_MINUS. Maybe someone more aware of keyboard mapping could tell if this is normal.

Apart from the wrong separator in lua_api.txt, the rest looks good to me.

rubenwardy commented:
> Number of fixed order elements begins to grow (size > position > anchor > no_prepend > options > real_coordinates). I'm not sure it's always relevant to fix that order (it may generate unnecessary errors). Hope in the furure we will have two main types of elements: "global" elements and "other" elements. "global" elements just having to be before any "other" elements but in any order.

I prefer the terms header and body, or head and body. To match both packets (they have header and payloads) and HTML (`<head>` and `<body>`).

v-rob commented:
> I get some exotic key mapping with my AZERTY keyboard. For example, if I press shift-4 (shift is necessary on Azerty to get the number instead of single quote), I get a KEY_OEM_7. For 6 and 8 I get KEY_MINUS. Maybe someone more aware of keyboard mapping could tell if this is normal.

Probably related to #302.

I agree with head vs. body elements; I'll keep it in mind in the future.  Also, allowing mouse moves to be filtered is a great idea; I'll implement now since it's really easy.

v-rob commented:
Note about SDL2's relation to this PR:  While SDL is liable to change some things, it doesn't necessarily prevent this PR from going forward.  Notably, Irrlicht's `KeyEvent::Char` field is absent from this PR, so that is no problem from a SDL2 standpoint.  Keycode names will be different in SDL2, but compatibility can be worked in if necessary.  Ideally, this PR would wait for SDL2, but it's in high demand.

v-rob commented:
Three changes:
* The keymap names feature was broken due to recent settings changes.  This is fixed.
* All the extraneous `control_down`, `shift_down`, etc were removed since they really are redundant.  Shift, control, and mouse button states can be tracked easily enough manually.
* The mouse events were split into three fields, e.g. `type="left_doubleclick"` has been split into `type="click", button="left", clicks=2`, which is easier to work with.

Sorry to change the PR at this stage, but I think it's better this way.

kilbith commented:
What's the status of this?

v-rob commented:
Awaiting review.

appgurueu commented:
Should this really be stuffed in `on_player_receive_fields`?

sfan5 commented on doc/keycodes.txt:
```diff
@@ -0,0 +1,162 @@
+that are on every or most keyboards.
+
+* `KEY_LBUTTON`          = Left mouse button
+* `KEY_RBUTTON`          = Right mouse button
```
I know this has been brought up before but with the Irrlicht 1.9 switch and possible(?) SDL integration coming I think that making this list of keys essentially part of our API is an awful idea.
Especially the IME keys, `KEY_OEM_2` and arcane keys like `KEY_XBUTTON1` (mouse input shouldn't be here anyway) are likely to go once someone sorts out the input system mess.

I'd solve this differently, specifically I'd identify keys by the character they produce and use reserved names for the rest.
So e.g.:
* printable ASCII space: character keys, always lowercase (example: `a`)
* `<Tab>`: Tab key
* `<Backspace>`: ...
* `<Return>`: ...
* `<Left>`, `<Right>`, `<Up>`, `<Down>`: direction keys

sfan5 replied:
Also there's something I don't see explained here: How would Shift+A be sent? Are held down keys sent repeatedly? (I hope not)

v-rob replied:
I don't like it either.  SDL doesn't even provide a certain way to do this: `SDL_GetKeyName` and `SDL_GetScancodeName` can give different results on different platforms (like `SDL_SCANCODE_LGUI` being `"Left GUI"` on Linux and `"Left Windows"` on Windows).

We have our own keycode names used in `KeyCode` for usage in the key menu, but it's meant for human readability.  We could modify that somewhat and base it off of SDL keycodes, only documenting the keys that overlap between SDL and Irrlicht for now, and then documenting the rest of the SDL specific keys later.

Shift is sent as a standalone key (`KEY_L/RSHIFT`), and so is A (`KEY_KEY_A`).  If a key is held down, only the most recently pressed key repeats.  So, holding down Shift and A would look like `KEY_LSHIFT` -> `KEY_KEY_A` -> `KEY_KEY_A` -> `KEY_KEY_A` -> ...  So, you have to manually track shift pressed/releases to detect a Shift+A.

Desour replied:
Mods shouldn't hardcode the keycodes anyway imo. (I hate it when games hardcode the z and x buttons.)

Testing on my machine showed that Shift isn't sent repeatedly, but A is. It would probably be better to send hold down keys only once.

v-rob replied:
Well, hardcoded keys is necessary.  Perhaps the most useful use-case of this PR is to have keyboard shortcuts, and those must be bound to specific keys.

Key repetition is a useful feature (like in text fields).  I like SDL's implementation, which has a `repeat` field that tells whether or not this current event is a repetition, allowing you to filter it out.

loffren174 replied:
I know that the names 
```
* `KEY_LWIN`             = Left Windows key (Microsoft Natural keyboard)
* `KEY_RWIN`             = Right Windows key (Natural keyboard)
```
..are already used in keycode.cpp, but exposing them as such to lua means we can't change the names afterwards as easy. The key has different names on different platforms unix, windows,apple, which are interpreted as the according equivalent anyway:
https://en.wikipedia.org/wiki/Meta_key
https://en.wikipedia.org/wiki/Windows_key
https://en.wikipedia.org/wiki/Command_key

If we're free to call it whatever we want, maybe 'meta key' is more neutral regarding to systems, i.e. equally unknown to both apple users and windows users, but easily guessable.  I know that on kde plasma the name 'meta key' is used, for example.  I didn't find anything on how this is called on android, and I never saw it on android systems' screen keyboards either. Still, given that most keyboards are not apple keyboards, but have a windows logo printed on them, it's also okay to call it that, I guess - especially if we already define the key to this name internally.

sfan5 commented on doc/lua_api.txt:
```diff
@@ -2181,12 +2181,27 @@ Elements
+    * mouse_event: Sets whether to have mouse events in the field `mouse_event`.
+      Possible values:
+        * `none` (default): Mouse events will not be sent.
+        * `all`: Sends all mouse events.
+        * `no_move`: All mouse events will be registered except for `move`; this drastically
+          reduces the amount of mouse events sent, improving performance.
```
Then, is there no event ratelimiting at all? That sounds like it'd lead to problems as soon as used anywhere except singleplayer/LAN.
Limiting these two one move event per 200ms would solve this but I'm sure it breaks some usecases people would like (like a paint application, do we want to support that?).

Handling these kinda of events on the server is doomed to always suffer from latency. Perhaps we should <i>wait for SSCSM</i>™ and make this API available to it then.

v-rob replied:
Are you suggesting to limit move events only to 200ms?  That sounds like it wouldn't lead to a lot of trouble (and a paint application sounds like abuse, formspec boxes would be too slow anyway).  Limiting other things like keys or mouse clicks to 200ms could cause trouble since they could be missed very easily just like `player:get_player_control()`.

Also, a (much nicer) event API in CSM will be necessary for the formspec replacement, so SSCSM will have it if it ever comes :)

sfan5 replied:
> Are you suggesting to limit move events only to 200ms? 

Yes, only move events. If desired this can even be improved by not delaying sending if the mouse just started/stopped moving.

v-rob commented:
> Should this really be stuffed in `on_player_receive_fields`?

Where else would it go?  That's where all formspec events go, and it would be weird to just have these go somewhere else.

Desour reviewed (commented):
Tested, works fine (apart from the stuff already discussed above).
Code also looks good.
:+1:

v-rob committed:
Add rate limiting to "move" events (one per 200 milliseconds)

appgurueu commented:
Why limit this to formspecs? Why not send all key & mouse events to the server? This would finally close #2731.

v-rob commented:
Because, quite simply, it was easier to do it this way at the time I made this.

I'm not going to send all key and mouse events to the server until I implement SDL2 as the event backend for Irrlicht.  Then, I'll make a full PR to send not only key and mouse, but joystick, touchscreen, and other events to the server.

sfan5 commented:
Are we making a VNC client or a game engine?
Latency is already sort of an issue with this and I question that shoving everything to the server is the right design. Rather I'd like to  see the ability for (SS)CSM to bind keys or related events and if modders want to they can still just forward everything to the server.

v-rob commented:
I'd love to see SSCSM have this ability and not the server.  Actually, I'd love to see SSCSM do anything.  I'd love to see SSCSM at all.  But I don't see it happening.  Better to give the server some sort of control than no one has control.  Note that I don't propose to give the server the same complexity of events as GUIs because it has different needs than a GUI does, so latency would be easier with simpler events.

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -4098,28 +4235,27 @@ enum ButtonEventType : u8
-			m_client->makeScreenshot();
+		if (m_use_key_event) {
+			current_keys_pending.pending_key_event = true;
+			current_keys_pending.key_event = event.KeyInput;
```
from my understanding, it looks like this only allows on key event to be queued at a time. It's possible for the OS to deliver more than one key press when the program polls for input. This is especially likely if the program's game loop has just frozen for a period

v-rob replied:
Oh yes.  I dealt with this problem in #10524, but not in this PR.  Will fix.

v-rob replied:
Actually, now that I'm actually looking at this again, this isn't a problem.  When `GUIFormSpecMenu` receives an event in `OnEvent()`, if it's a key or mouse event, it dumps it into `current_keys_pending` and ships it off to `acceptInput()`, which then builds the `fields` table that you get in `minetest.register_on_player_receive_fields()` and sends it to the server.  Hence, it's a direct line from `OnEvent()` to the server, and `OnEvent()` can get called as many times a frame as Irrlicht so pleases.

rubenwardy reviewed (commented):
`minetest.dmg` shouldn't be included in this PR. I also agree that we should decide on an API, including key names, taking a switch to SDL into account

rubenwardy commented:
>  Latency is already sort of an issue with this and I question that shoving everything to the server is the right design. Rather I'd like to see the ability for (SS)CSM to bind keys or related events and if modders want to they can still just forward everything to the server.

I agree that we should add support for custom mod bindings. The mod could register an action and it would show in the key change menu. However, having access to mouse and keyboard events is easier and still useful. My main concern with this is deciding on an API that makes sense and doesn't trap us in the future (although, with the formspec replacement this may not be an issue)

v-rob commented:
I don't know how `minetest.dmp` got in (or what it actually even is), so I'll remove it.  As for key names, they're still on the plan.  I'll make time for these PRs soon.

v-rob committed:
Rewrite KeyPress; attempt to be Irrlicht-independent

v-rob commented:
Finally got around to doing this.  The main change is that `KeyPress` has been entirely rewritten.  `KeyPress` had a lot of extraneous info for no use at all.  It was possible to bind a key to `7`, but also to `&`, even when they share the same key, by doing `Shift+7`.  Of course, you can't bind other keys to e.g. `x` and `Shift+x` for `X`, so it's just plain wrong to have that behaviour.  This is also the root cause of the failure of #10419 -- `&` has no Irrlicht keycode.

Hence, `KeyPress` is now just a 1:1 Irrlicht keycode <-> string mapping.  Theoretically, `KeyPress` could be removed altogether, but I decided it's probably best to make that decision elsewhere, if at all.  Due to the simplicity of `KeyPress` now, I removed the unit test -- there's simply nothing useful to test anymore.  The main thing is to test with physical keyboards and different platforms, and test with every single key on your keyboard and make sure they give an appropriate keycode.  `&` and suchlike should be impossible.  Hopefully, the Irrlicht OEM keys are stable enough across keyboards/platforms (SDL gives them concrete names instead of OEM #), but I'm not entirely certain--testing is necessary to be certain.  You can view the new list of keycode names in `doc/keycodes.txt`: they're tentative, and I'm open to changing them around.

Finally, I painstakingly went though all of the SDL scancodes and Irrlicht keycodes together, so the only ones in Minetest are the codes shared between them, which ought to be general enough for any input backend.  I also compiled a list of SDL scancodes not supported by Irrlicht and vice-versa, and a list of names to change in `KeyPress` to transfer from Irrlicht to SDL, for possible future reference if we ever get there.

There's one exception to all of this: `KEY_L/M/RBUTTON` is still in `KeyPress` since that's used in keybindings, but it will never appear in `key_event`.

<details><summary><b>SDL/Irrlicht Scancode/Keycode relationship lists:</b></summary>

SDL scancodes that are not shared with Irrlicht:
 * SDL_SCANCODE_AC_BACK
 * SDL_SCANCODE_AC_BOOKMARKS
 * SDL_SCANCODE_AC_FORWAD
 * SDL_SCANCODE_AC_HOME
 * SDL_SCANCODE_AC_REFRESH
 * SDL_SCANCODE_AC_SEARCH
 * SDL_SCANCODE_AC_STOP
 * SDL_SCANCODE_AGAIN
 * SDL_SCANCODE_ALTERASE
 * SDL_SCANCODE_AUDIOMUTE
 * SDL_SCANCODE_AUDIONEXT
 * SDL_SCANCODE_AUDIOPLAY
 * SDL_SCANCODE_AUDIOPREV
 * SDL_SCANCODE_AUDIOSTOP
 * SDL_SCANCODE_BRIGHTNESSDOWN
 * SDL_SCANCODE_BRIGHTNESSUP
 * SDL_SCANCODE_CALCULATOR
 * SDL_SCANCODE_CLEARAGAIN
 * SDL_SCANCODE_COMPUTER
 * SDL_SCANCODE_COPY
 * SDL_SCANCODE_CURRENCYSUBUNIT
 * SDL_SCANCODE_CURRENCYUNIT
 * SDL_SCANCODE_CUT
 * SDL_SCANCODE_DECIMALSEPARATOR
 * SDL_SCANCODE_DISPLAYSWITCH
 * SDL_SCANCODE_EJECT
 * SDL_SCANCODE_FIND
 * SDL_SCANCODE_KBDILLUMDOWN
 * SDL_SCANCODE_KBDILLUMTOGGLE
 * SDL_SCANCODE_KBDILLUMUP
 * SDL_SCANCODE_KP_00
 * SDL_SCANCODE_KP_000
 * SDL_SCANCODE_KP_A
 * SDL_SCANCODE_KP_AMPERSAND
 * SDL_SCANCODE_KP_AT
 * SDL_SCANCODE_KP_B
 * SDL_SCANCODE_KP_BACKSPACE
 * SDL_SCANCODE_KP_BINARY
 * SDL_SCANCODE_KP_C
 * SDL_SCANCODE_KP_CLEAR
 * SDL_SCANCODE_KP_CLEARENTRY
 * SDL_SCANCODE_KP_COLON
 * SDL_SCANCODE_KP_COMMA
 * SDL_SCANCODE_KP_D
 * SDL_SCANCODE_KP_DBLAMPERSAND
 * SDL_SCANCODE_KP_DBLVERTICALBAR
 * SDL_SCANCODE_KP_E
 * SDL_SCANCODE_KP_ENTER
 * SDL_SCANCODE_KP_EQUALS
 * SDL_SCANCODE_KP_EQUALSAS400
 * SDL_SCANCODE_KP_EXCLAM
 * SDL_SCANCODE_KP_F
 * SDL_SCANCODE_KP_GREATER
 * SDL_SCANCODE_KP_HASH
 * SDL_SCANCODE_KP_HEXADECIMAL
 * SDL_SCANCODE_KP_LEFTBRACE
 * SDL_SCANCODE_KP_LEFTPAREN
 * SDL_SCANCODE_KP_LESS
 * SDL_SCANCODE_KP_MEMADD
 * SDL_SCANCODE_KP_MEMCLEAR
 * SDL_SCANCODE_KP_MEMDIVIDE
 * SDL_SCANCODE_KP_MEMMULTIPLY
 * SDL_SCANCODE_KP_MEMRECALL
 * SDL_SCANCODE_KP_MEMSTORE
 * SDL_SCANCODE_KP_MEMSUBTRACT
 * SDL_SCANCODE_KP_OCTAL
 * SDL_SCANCODE_KP_PERCENT
 * SDL_SCANCODE_KP_PERIOD
 * SDL_SCANCODE_KP_PLUSMINUS
 * SDL_SCANCODE_KP_POWER
 * SDL_SCANCODE_KP_RIGHTBRACE
 * SDL_SCANCODE_KP_RIGHTPAREN
 * SDL_SCANCODE_KP_SPACE
 * SDL_SCANCODE_KP_TAB
 * SDL_SCANCODE_KP_VERTICALBAR
 * SDL_SCANCODE_KP_XOR
 * SDL_SCANCODE_MAIL
 * SDL_SCANCODE_MEDIASELECT
 * SDL_SCANCODE_MUTE
 * SDL_SCANCODE_OPER
 * SDL_SCANCODE_OUT
 * SDL_SCANCODE_PASTE
 * SDL_SCANCODE_POWER
 * SDL_SCANCODE_PRIOR
 * SDL_SCANCODE_RETURN2
 * SDL_SCANCODE_STOP
 * SDL_SCANCODE_SYSREQ
 * SDL_SCANCODE_THOUSANDSSEPARATOR
 * SDL_SCANCODE_UNDO
 * SDL_SCANCODE_VOLUMEDOWN
 * SDL_SCANCODE_VOLUMEUP
 * SDL_SCANCODE_WWW
 * SDL_SCANCODE_INTERNATIONAL1
 * SDL_SCANCODE_INTERNATIONAL2
 * SDL_SCANCODE_INTERNATIONAL3
 * SDL_SCANCODE_INTERNATIONAL4
 * SDL_SCANCODE_INTERNATIONAL5
 * SDL_SCANCODE_INTERNATIONAL6
 * SDL_SCANCODE_INTERNATIONAL7
 * SDL_SCANCODE_INTERNATIONAL8
 * SDL_SCANCODE_INTERNATIONAL9
 * SDL_SCANCODE_LANG3
 * SDL_SCANCODE_LANG4
 * SDL_SCANCODE_LANG5
 * SDL_SCANCODE_LANG6
 * SDL_SCANCODE_LANG7
 * SDL_SCANCODE_LANG8
 * SDL_SCANCODE_LANG9
 * SDL_SCANCODE_LOCKINGCAPSLOCK
 * SDL_SCANCODE_LOCKINGNUMLOCK
 * SDL_SCANCODE_LOCKINGSCROLLLOCK
 * SDL_SCANCODE_NONUSBACKSLASH
 * SDL_SCANCODE_NONUSHASH

SDL keycodes that are not scancodes.  Generally, Minetest should avoid these:
 * SDLK_AMPERSAND
 * SDLK_ASTERISK
 * SDLK_AT
 * SDLK_CARET
 * SDLK_COLON
 * SDLK_DOLLAR
 * SDLK_EXCLAIM
 * SDLK_GREATER
 * SDLK_HASH
 * SDLK_LEFTPAREN
 * SDLK_LESS
 * SDLK_PERCENT
 * SDLK_PLUS
 * SDLK_QUESTION
 * SDLK_QUOTEDBL
 * SDLK_RIGHTPAREN
 * SDLK_UNDERSCORE

Irrlicht keycodes that SDL does not have (note that Minetest uses L/R/MBUTTON for mouse bindings):
 * KEY_LBUTTON
 * KEY_RBUTTON
 * KEY_MBUTTON
 * KEY_XBUTTON1
 * KEY_XBUTTON2
 * KEY_KANA
 * KEY_HANGUEL
 * KEY_HANGUL
 * KEY_JUNJA
 * KEY_FINAL
 * KEY_HANJA
 * KEY_KANJI
 * KEY_CONVERT
 * KEY_NONCONVERT
 * KEY_ACCEPT
 * KEY_PRINT
 * KEY_OEM_8
 * KEY_OEM_AX
 * KEY_OEM_102
 * KEY_ATTN
 * KEY_EREOF
 * KEY_PLAY
 * KEY_ZOOM
 * KEY_PA1
 * KEY_OEM_CLEAR

In the Irrlicht keycode <-> name map in `keycode.cpp`, this is a list of names to change to convert them from Irrlicht keycode names to SDL scancode names when they differ:
 * OEM_7 -- APOSTROPHE
 * APPS -- APPLICATION
 * BACK -- BACKSPACE
 * CAPITAL -- CAPSLOCK
 * PLUS -- EQUALS
 * EXECUT -- EXECUTE
 * OEM_3 -- GRAVE
 * NUMPAD0 -- KP_0
 * NUMPAD1 -- KP_1
 * NUMPAD2 -- KP_2
 * NUMPAD3 -- KP_3
 * NUMPAD4 -- KP_4
 * NUMPAD5 -- KP_5
 * NUMPAD6 -- KP_6
 * NUMPAD7 -- KP_7
 * NUMPAD8 -- KP_8
 * NUMPAD9 -- KP_9
 * SUBTRACT -- KP_MINUS
 * MULTIPLY -- KP_MULTIPLY
 * DIVIDE -- KP_DIVIDE
 * DECIMAL -- KP_PERIOD
 * ADD -- KP_PLUS
 * LMENU -- LALT
 * LCONTROL -- LCTRL
 * OEM_4 -- LEFTBRACKET
 * LWIN -- LGUI
 * MODECHANGE -- MODE
 * NUMLOCK -- NUMLOCKCLEAR
 * NEXT -- PAGEDOWN
 * PRIOR -- PAGEUP
 * SNAPSHOT -- PRINTSCREEN
 * RMENU -- RALT
 * RCONTROL -- RCTRL
 * RWIN -- RGUI
 * OEM_6 -- RIGHTBRACKET
 * SCROLL -- SCROLLLOCK
 * OEM_1 -- SEMICOLON
 * OEM_2 -- SLASH
</details>

v-rob committed:
Bump formspec version to 6

v-rob commented:
Also: bumped formspec version to 5 since this adds a new element (`options[]`)

kilbith commented:
If you want it to be extensively tested, your best bet is to merge it so I can effectively give my feedbacks out, otherwise I won't bother test this PR. You need to push your PRs more "aggressively".

kilbith commented:
Please rebase.

@Df458 Please test.

Df458 commented:
@kilbith I try to avoid testing anything that isn't in a state where it can be merged. If this gets rebased, I'll give it a pass then.

Df458 commented:
I'm still thinking about how to best go about testing this. My current thought is to develop some formspec-based test mods designed for the new behavior, and hopefully let others help verify things like how they perform in a server environment. Current test ideas:
- Tabbed UI with hotkey tab-switching
- Some sort of "minigame"-style mod with appearing/disappearing images for testing click events on elements
- Some sort of "viewer"-style mod that uses drag panning & scrollwheel zoom
- A formspec that attempts to mimic certain player actions using their mappings

This is probably going to take some time to implement and verify these, I'm open to other ideas for testing in the meantime.

appgurueu commented:
> This is probably going to take some time to implement and verify these, I'm open to other ideas for testing in the meantime.

Just make this available to modders in an "experimental" build :)

appgurueu commented on builtin/common/misc_helpers.lua:
```diff
@@ -400,6 +400,80 @@ function core.explode_scrollbar_event(evt)
 	return retval
 end
 
+--------------------------------------------------------------------------------
```
Do we really need this clutter? Feel free to insert some blank lines, but these separators are BS and add absolutely no value.

rubenwardy replied:
This matches the rest of builtin, out-of-scope for this PR (but I agree, it should just be two blank lines)

v-rob replied:
Yeah, just for consistency with builtin--I don't like it either.

appgurueu commented on builtin/common/misc_helpers.lua:
```diff
@@ -400,6 +400,80 @@ function core.explode_scrollbar_event(evt)
+	return ret
+end
+
+--------------------------------------------------------------------------------
```
Same thing here.

appgurueu commented on builtin/common/misc_helpers.lua:
```diff
@@ -400,6 +400,80 @@ function core.explode_scrollbar_event(evt)
+--------------------------------------------------------------------------------
+function core.explode_key_event(evt)
+	if evt == nil then
+		return nil
```
`return nil` or just `return`? The latter is cleaner IMO (has to be documented appropriately though).

appgurueu commented on builtin/common/misc_helpers.lua:
```diff
@@ -400,6 +400,80 @@ function core.explode_scrollbar_event(evt)
+		return nil
+	end
+
+	local parts = evt:split(";", true)
```
At this point I'm wondering why you don't just send events as JSON.

v-rob replied:
I tried that before.  `minetest.parse_json()` prints bunches of errors on invalid JSON or something like that, I forget the exact reason why I eventually decided against it.

appgurueu replied:
`minetest.parse_json` is indeed crap [as documented here](https://github.com/minetest/minetest_docs/pull/13) but should suffice for this narrow use case

v-rob commented:
So, after a lengthy IRC discussion, we have discovered:
* Keycodes are inconsistent pieces of garbage.  They're no good for WASD-style stuff, and they're no good for keyboard shortcuts either since they can change between platforms.  Irrlicht already tries its best effort at making these work, it seems, but it's not good enough.
* Having scancodes ("X and Y" physical key coordinates is the intuitive way of thinking about it) is a good idea for WASD-style stuff, but Irrlicht doesn't have this natively
* Instead of keycodes, it's probably a better idea to return the character for the key.  This isn't _perfect_ in terms of Unicode, and definitely shouldn't be used for text input fields, but will allow keyboard shortcuts.  You can check the scancode for shift and stuff, and the character for the keyboard shortcut type stuff.

In other words, this PR is totally unsuitable in its current form for processing anything other than shift, control, spacebar, enter, etc. in terms of the keyboard.  It will possibly need Irrlicht changes to expose scancodes.

I do worry, however, that if SDL2 is ever realized, that its characters will be inconsistent with Irrlicht's `Char` field, and there will be backwards compatibility muck.

v-rob commented:
The idea behind "character" is the single Unicode codepoint associated with that key.  Not like text input (which can, as you said, use multiple characters), but just the single immediate character like `a` for the A key, `A` for Shift+A, plus special thing like `<Shift>` for shift, and this would automatically support things like, say, `ö` keys.  At the same time, it's less than ideal because the shift key modifies things.  I'd definitely prefer real keysyms that work properly across keyboard layouts.

(Unfortunately, even if we exposed scancodes or keysyms only, people would still probably use them for text input on slow bad-with-latency homemade input fields that obviously don't work with Unicode.  It's hard to avoid.)

I don't want to jump to conclusions; it's possible that Minetest is doing something funky that got some of the mangled results we had earlier, or some may be quirks of erlehmann's keyboard that really don't extend to the vast majority of keyboards.  I like sfan5's suggestion of really making a program to test this stuff out rather than jumping to any conclusions and finding that they're wrong later.

Zughy commented:
@v-rob I'm closing this since it's been more than a month with the "Action / Change needed" label, but feel free to reopen it if you still want to work on it (I've also tried to contact you on Matrix to tell you)
